src/engagement_db_to_analysis/code_imputation_functions.py

```diff
@@ -333,12 +329,111 @@ def _impute_true_missing(user, column_traced_data_iterable, analysis_dataset_con
+                normal_code = code
+
+            if normal_code.code_id != code.code_id:
+                return True
+
+    return False
+
+
+def _get_control_and_meta_labels(column_traced_data, column_config):
+    """
+    :param column_traced_data: Column-view traced data to get the control and meta labels from.
+    :type column_traced_data: core_data_modules.traced_data.TracedData
+    :param column_config: Configuration for the column to get the labels from.
+    :type column_config: core_data_modules.analysis.analysis_utils.AnalysisConfiguration
+    :return: The labels in `column_traced_data` that have code_type CONTROL or META under `column_config`, serialized.
+    :rtype: list of dict
+    """
+    control_and_meta_labels = []
+    column_labels = column_traced_data[column_config.coded_field]
+    for label in column_labels:
+        code = column_config.code_scheme.get_code_with_code_id(label["CodeID"])
+        if code.code_type == CodeTypes.CONTROL or code.code_type == CodeTypes.META:
+            control_and_meta_labels.append(label)
+    return control_and_meta_labels
+
+
+def _impute_nic_demogs(user, column_traced_data_iterable, analysis_dataset_configs):
+    """
+    Imputes NOT_INTERNALLY_CONSISTENT labels on the demographic columns of column-view datasets.
+
+    NOT_INTERNALLY_CONSISTENT labels are applied to demographics where there are multiple, conflicting normal labels.
+    For example:
+     - If we have multiple conflicting normal labels e.g. "20" and "22", this would be converted to
+       NOT_INTERNALLY_CONSISTENT.
+     - If we have a a single normal age label and some meta/control labels, no imputation is performed.
+     - If we have multiple normal age labels but these have the same code id, e.g. for the messages "21" and "I'm 21",
+       no imputation is performed.
+     - If we have multiple conflicting normal labels, e.g. "20" and "22", as well as some meta/control labels, the
+       normal labels will be replaced with NOT_INTERNALLY_CONSISTENT and the meta/control labels will be kept untouched.
+
+    :param user: Identifier of user running the pipeline.
+    :type user: str
+    :param column_traced_data_iterable: Column-view traced data objects to apply the impute function to.
+    :type column_traced_data_iterable: iterable of core_data_modules.traced_data.TracedData
+    :param analysis_dataset_configs: Analysis dataset configurations for the imputation.
+    :type analysis_dataset_configs: pipeline_config.analysis_configs.dataset_configurations
+    """
+    imputed_codes = 0
+    log.info(f"Imputing {Codes.NOT_INTERNALLY_CONSISTENT} codes...")
+
+    demog_column_configs = _get_demog_column_configs(analysis_dataset_configs)
+    for td in column_traced_data_iterable:
+        for column_config in demog_column_configs:
+            if _demog_has_conflicting_normal_labels(td, column_config):
+                # Replace the conflicting normal labels with a NOT_INTERNALLY_CONSISTENT label,
```

Comment:
Should we entirely replace or impute NIC, retain Meta/Control codes, and keep conflicting labels history?

Reply:
discussed offline, will keep as-is as we're not working with coda style datasets here so don't need to mix history with data